### DISCUSSION TITLE
Getting all entities within a specified bounding box

### DISCUSSION THREAD

**Discussion by c-chaitanya** [mozman/ezdxf#600] [Q&A] [answered]
Say i was given four(rectangle) or five(polygon) coordinates of a dxf  can I get all entities that lie within those cordinates?

I have this code where I can get the starting and ending points of each entity in the dxf
```
msp = doc.modelspace()
cache = bbox.Cache()
for e in msp:
    bbox_ = bbox.multi_flat(msp.query(e.dxftype()), cache=cache)
    print(*bbox_ , e.dxftype())
```
I was checking for a way if I could get all entities that lie within a specified bounding box.

**Comment by mozman** [comment]
# Inside Check

Given two bounding boxes `box_a` and `box_b`, test if `box_a` is completely inside of `box_b`:

```Python
# for 3D BoundingBox(), the default in most cases:
print(box_b.all_inside(box_a.cube_vertices()))

# for 2D BoundingBox2d():
print(box_b.all_inside(box_a.rect_vertices()))
```
# Intersection Check

```Python
print(box_b.intersect(box_a))  
```

**Reply by c-chaitanya** [reply]
Thanks for the reply @mozman . But How do I get all entities(lines, circle ..etc) within that bounding box?

**Reply by mozman** [reply]
I don't understand what your code should achieve, but the basics are simple:

1. calculate the test bounding box: `box_b`
2. calculate the bounding box for each entity: `box_a`
3. do the test: `box_b.inside(box_a.extmin) and box_b.inside(box_a.extmax)`

**Reply by mozman** [reply]
additional hints:
```Python
b = bbox.extents([e0, e1, e2, ...])  # bounding box for multiple entities
b = bbox.extents([e0])  # bounding box for a single entity
```

**Comment by mozman** [comment]
Better, faster solution and works for 2d and 3d bounding box the same way:

```Python
print(box_b.inside(box_a.extmin) and box_b.inside(box_a.extmax))
```

**Reply by c-chaitanya** [reply]
Hi @mozman  ,

I am confused about how bounding boxes are to be represented. In My case I'm given this polygon(Bounding Box) ```[(495, 585), (547, 585), (547, 537), (495, 537), (495, 585)]```. I am to find all entities that lie within this polygon. Specifying Bounding boxes as ```bbox_b = bbox.extents([(495, 585), (547, 585), (547, 537), (495, 537), (495, 585)]) ``` doesn't seem to work.

Here is my code for which I get the error ```AttributeError: 'tuple' object has no attribute 'dxftype'``` How do I represent my bounding boxes?
```
lst = ['ARC','CIRCLE','HATCH','INSERT','LINE','LWPOLYLINE','MTEXT']
bbox_b = bbox.extents([(495, 585), (547, 585), (547, 537), (495, 537), (495, 585)])
cache = bbox.Cache()
for entity in lst:
    bbox_ = bbox.multi_flat(msp.query(entity), cache=cache)
    for bbox_a in bbox_:
        bbox_a = bbox.extents(bbox_a)
        print(bbox_b.inside(bbox_a.extmin) and bbox_b.inside(bbox_a.extmax))
```

Any Help is greatly appreciated.

**Comment by mozman** [comment, marked as answer]
The `bbox.extends()` method calculates bounding boxes of entities: https://ezdxf.mozman.at/docs/bbox.html

The result is a bounding box class:  https://ezdxf.mozman.at/docs/math.html#boundingbox

```Python
from ezdxf.math import BoundingBox

dxf_types = ['ARC','CIRCLE','HATCH','INSERT','LINE','LWPOLYLINE','MTEXT']
search_box = BoundingBox([(495, 585), (547, 585), (547, 537), (495, 537), (495, 585)])
cache = bbox.Cache()
for dxf_type in dxf_types:
    entity_boxes = bbox.multi_flat(msp.query(dxf_type), cache=cache)
    for entity_box in entity_boxes:
        print(search_box.inside(entity_box.extmin) and search_box.inside(entity_box.extmax))

print(cache)  # if there are very few hits and you don't reuse the cache for further calculations, you don't need a cache!

```
A cache is nothing magical and no one seems to read this part of the documentation: https://ezdxf.mozman.at/docs/bbox.html#caching-strategies

**Comment by c-chaitanya** [comment]
Thank You @mozman for the help. Its really a great and helpful library you have built here.
